File: packages/db-schema/src/schema/calendar-external-events.ts
import { sql } from 'drizzle-orm'
import { integer, sqliteTable, text, index, uniqueIndex } from 'drizzle-orm/sqlite-core'
import type { VectorClock } from '@memry/contracts/sync-api'
import { calendarSources } from './calendar-sources.ts'
import type {
  CalendarAttendee,
  CalendarReminders,
  CalendarConferenceData,
  CalendarVisibility
} from './calendar-events.ts'

export type CalendarExternalEventStatus = 'confirmed' | 'tentative' | 'cancelled'

export const calendarExternalEvents = sqliteTable(
  'calendar_external_events',
  {
    id: text('id').primaryKey(),
    sourceId: text('source_id')
      .notNull()
      .references(() => calendarSources.id, { onDelete: 'cascade' }),
    remoteEventId: text('remote_event_id').notNull(),
    remoteEtag: text('remote_etag'),
    remoteUpdatedAt: text('remote_updated_at'),
    title: text('title').notNull(),
    description: text('description'),
    location: text('location'),
    startAt: text('start_at').notNull(),
    endAt: text('end_at'),
    timezone: text('timezone'),
    isAllDay: integer('is_all_day', { mode: 'boolean' }).notNull().default(false),
    status: text('status').$type<CalendarExternalEventStatus>().notNull().default('confirmed'),
    recurrenceRule: text('recurrence_rule', { mode: 'json' }).$type<Record<string, unknown> | null>(),
    attendees: text('attendees', { mode: 'json' }).$type<CalendarAttendee[] | null>(),
    reminders: text('reminders', { mode: 'json' }).$type<CalendarReminders | null>(),
    visibility: text('visibility').$type<CalendarVisibility | null>(),
    colorId: text('color_id'),
    conferenceData: text('conference_data', { mode: 'json' }).$type<CalendarConferenceData | null>(),
    rawPayload: text('raw_payload', { mode: 'json' }).$type<Record<string, unknown> | null>(),
    archivedAt: text('archived_at'),
    clock: text('clock', { mode: 'json' }).$type<VectorClock>(),
    syncedAt: text('synced_at'),
    createdAt: text('created_at')
      .notNull()
      .default(sql`(strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))`),
    modifiedAt: text('modified_at')
      .notNull()
      .default(sql`(strftime('%Y-%m-%dT%H:%M:%fZ', 'now'))`)
  },
  (table) => [
    uniqueIndex('idx_calendar_external_events_source_remote').on(table.sourceId, table.remoteEventId),
    index('idx_calendar_external_events_start_at').on(table.startAt),
    index('idx_calendar_external_events_archived_at').on(table.archivedAt)
  ]
)

export type CalendarExternalEvent = typeof calendarExternalEvents.$inferSelect
export type NewCalendarExternalEvent = typeof calendarExternalEvents.$inferInsert

File: apps/desktop-tauri/src-tauri/src/db/calendar_external_events.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, specta::Type)]
#[serde(rename_all = "camelCase")]
pub struct CalendarExternalEvent {
    pub id: String,
    pub source_id: String,
    pub remote_event_id: String,
    pub remote_etag: Option<String>,
    pub remote_updated_at: Option<String>,
    pub title: String,
    pub description: Option<String>,
    pub location: Option<String>,
    pub start_at: String,
    pub end_at: Option<String>,
    pub timezone: Option<String>,
    pub is_all_day: bool,
    pub status: String,
    pub recurrence_rule: Option<String>,
    pub raw_payload: Option<String>,
    pub archived_at: Option<String>,
    pub clock: Option<String>,
    pub synced_at: Option<String>,
    pub created_at: String,
    pub modified_at: String,
    pub attendees: Option<String>,
    pub reminders: Option<String>,
    pub visibility: Option<String>,
    pub color_id: Option<String>,
    pub conference_data: Option<String>,
}

impl CalendarExternalEvent {
    pub fn from_row(row: &rusqlite::Row<'_>) -> rusqlite::Result<Self> {
        Ok(Self {
            id: row.get("id")?,
            source_id: row.get("source_id")?,
            remote_event_id: row.get("remote_event_id")?,
            remote_etag: row.get("remote_etag")?,
            remote_updated_at: row.get("remote_updated_at")?,
            title: row.get("title")?,
            description: row.get("description")?,
            location: row.get("location")?,
            start_at: row.get("start_at")?,
            end_at: row.get("end_at")?,
            timezone: row.get("timezone")?,
            is_all_day: row.get::<_, i64>("is_all_day")? != 0,
            status: row.get("status")?,
            recurrence_rule: row.get("recurrence_rule")?,
            raw_payload: row.get("raw_payload")?,
            archived_at: row.get("archived_at")?,
            clock: row.get("clock")?,
            synced_at: row.get("synced_at")?,
            created_at: row.get("created_at")?,
            modified_at: row.get("modified_at")?,
            attendees: row.get("attendees")?,
            reminders: row.get("reminders")?,
            visibility: row.get("visibility")?,
            color_id: row.get("color_id")?,
            conference_data: row.get("conference_data")?,
        })
    }
}
